src/circuit/gadgets/pointer.rs

```diff
@@ -254,9 +255,9 @@ impl<F: LurkField> AllocatedPtr<F> {
         store.fetch_scalar(&scalar_ptr)
     }
 
-    pub fn scalar_ptr(&self, store: &Store<F>) -> Option<ScalarPtr<F>> {
+    pub fn scalar_ptr(&self, _store: &Store<F>) -> Option<ZExprPtr<F>> {
         let (tag, value) = (self.tag.get_value()?, self.hash.get_value()?);
-        store.scalar_from_parts(tag, value)
+        Some(ZPtr(ExprTag::from_field(&tag)?, value))
```

Comment:
This is not a clear improvement. Can we add back a convenience function equivalent to `scalar_from_parts()`?

Reply:
This is how `scalar_from_parts` was implemented:

```
pub fn scalar_from_parts(&self, tag: F, value: F) -> Option<ScalarPtr<F>> {
  let Some(e_tag) = ExprTag::from_field(&tag) else { return None };
  let scalar_ptr = ScalarPtr::from_parts(e_tag, value);
    self.scalar_ptr_map
      .contains_key(&scalar_ptr)
      .then_some(scalar_ptr)
    }
```

If we want we could implement an equivalent:

```
pub fn zptr_from_parts(&self, tag: F, value: F) -> Option<ScalarPtr<F>> {
  Some(ZPtr(ExprTag::from_field(&tag)?, value))
}
```

But this is pretty terse, so imo inlining actually improves the legibility of the code (i.e. we pay a dozen or so extra characters to save a mental stack slot and an editor jump).

That said, one reason to keep a helper function is if we do in fact need the `.contains_key` check that the `ZPtr` is actually in the `Store's` z_expr_ptr_map. I'm completely not sure we do need it (given that no test seems to require it). It also feels a little awkward to be overloading the `None` with a failure from tag conversion and `.contains_key`), so maybe we should do something like:

```rust
pub fn zptr_from_parts(&self, tag: F, value: F) -> Result<ScalarPtr<F>, Error<F>> {
  let tag = ExprTag::from_field(&tag).ok_or(Error("tag error"))
  let zptr = ZPtr(tag, value);
  if self.z_expr_ptr_map.contains_key(z_ptr) {
    Ok(zptr)
  } else {
    Error("uncached zptr")
  }
}
```

with `store::Error` (and then eventually with an actual error Enum in the future). Wdyt?

Reply:
Okay, just tried it and it seems to work (with an `.ok()` to get the Option back), so I've added `Store::z_expr_ptr_from_parts` and `Store::z_cont_ptr_from_parts`